File: Python/ByTier/F/2132_F_Rada_and_the_Chamomile_Valley.py
# Problem: CF 2132 F - Rada and the Chamomile Valley
# https://codeforces.com/contest/2132/problem/F

"""
F. Rada and the Chamomile Valley
Purpose:
This code determines which lanes (edges) in a connected graph are "forced" — meaning that all paths from node 1 to node n must use at least one of these lanes. The algorithm identifies bridge edges and then processes queries to determine the nearest forced edge for each given node.

Algorithms/Techniques Used:
- Bridge detection using Tarjan's algorithm (DFS with discovery time and lowest reachable vertex)
- Graph traversal to find bridges
- Shortest path BFS-style algorithm to assign distances and best edge indices

Time Complexity: O(n + m + q) per test case, where n is number of nodes, m is number of edges, and q is number of queries.
Space Complexity: O(n + m) for storing the graph, discovery times, low values, parent relationships, etc.

"""

import sys
from collections import deque

input = sys.stdin.readline


def solve():
    data = sys.stdin.buffer.read().split()
    it = iter(data)
    t = int(next(it))  # Read number of test cases
    out_lines = []
    for _case in range(t):
        n = int(next(it))   # Number of houses (nodes)
        m = int(next(it))   # Number of lanes (edges)
        U = [0] * (m + 1)   # u endpoints of edges (1-indexed)
        V = [0] * (m + 1)   # v endpoints of edges (1-indexed)
        g = [[] for _ in range(n + 1)]  # Adjacency list representation
        for i in range(1, m + 1):
            u = int(next(it))   # Read edge's start node
            v = int(next(it))   # Read edge's end node
            U[i] = u
            V[i] = v
            g[u].append((v, i))   # Add edge to adjacency list (node, id)
            g[v].append((u, i))   # Bidirectional graph

        q = int(next(it))     # Number of queries
        C = [int(next(it)) for __ in range(q)]  # Query nodes

        # Tarjan’s bridge-finding algorithm initialization
        tin = [0] * (n + 1)       # Discovery time for each node
        low = [0] * (n + 1)       # Lowest reachable discovery time of node
        tout = [0] * (n + 1)      # Finish time
        parent_edge = [0] * (n + 1)  # Parent edge id for each node
        parent = [0] * (n + 1)      # Parent node in DFS tree
        is_bridge = [False] * (m + 1)  # Tracks if an edge is a bridge
        timer = 0
        start = 1     # Start vertex for DFS
        stack = []
        stack.append((start, 0))   # Stack holds (node, next neighbor index)
        parent[start] = 0
        parent_edge[start] = 0

        # Running Tarjan’s algorithm to find all bridges
        while stack:
            u, idx = stack[-1]    # Peek at top of stack
            if tin[u] == 0:       # Node not yet discovered
                timer += 1
                tin[u] = low[u] = timer

            if idx < len(g[u]):   # If there are unvisited neighbors
                v, eid = g[u][idx]    # Get node and edge id
                stack[-1] = (u, idx + 1)  # Update to next neighbor in stack
                if eid == parent_edge[u]:  # Skip parent edge (avoid backtracking)
                    continue
                if tin[v] == 0:    # If neighbor not visited yet
                    parent[v] = u              # Set parent
                    parent_edge[v] = eid       # Set parent edge id
                    stack.append((v, 0))       # Push to DFS stack
                else:
                    # Back edge found - relax lowest reachable time
                    if tin[v] < tin[u]:
                        if tin[v] < low[u]:
                            low[u] = tin[v]
            else:      # All neighbors processed for node u
                if parent[u] != 0:
                    p = parent[u]
                    pe = parent_edge[u]
                    # Check if u -- pe -- p is a bridge
                    if low[u] > tin[p]:
                        is_bridge[pe] = True
                    # Propagate lowest value up the tree
                    if low[u] < low[p]:
                        low[p] = low[u]
                tout[u] = timer
                stack.pop()

        # Function to check if x belongs to subtree rooted at rootchild
        def in_subtree(x, rootchild):
            return tin[rootchild] <= tin[x] <= tout[rootchild]

        # Identify forced edges — bridges that separate source from destination components
        forced = [False] * (m + 1)
        s = 1      # Source node
        tnode = n  # Target node
        for v in range(1, n + 1):
            eid = parent_edge[v]
            if eid == 0:
                continue
            if not is_bridge[eid]:
                continue
            # Is the edge "forced"? i.e., separates s/t from each other through this bridge
            if in_subtree(s, v) ^ in_subtree(tnode, v):
                forced[eid] = True

        # Collect all forced edges
        forced_edges = [i for i in range(1, m + 1) if forced[i]]
        if not forced_edges:
            out_lines.append(" ".join(["-1"] * q))
            continue

        # Assign shortest distances and best (smallest edge id) to each node via BFS-like traversal
        INF = 10**18
        dist = [INF] * (n + 1)
        best = [10**18] * (n + 1)
        dq = deque()
        
        # Initialize with forced edges
        for eid in forced_edges:
            a = U[eid]
            b = V[eid]
            # Update distance and best edge if current is better than previous
            if dist[a] > 0 or (dist[a] == 0 and eid < best[a]):
                dist[a] = 0
                best[a] = eid
                dq.append(a)
            if dist[b] > 0 or (dist[b] == 0 and eid < best[b]):
                dist[b] = 0
                best[b] = eid
                dq.append(b)

        # BFS-style update of distances and optimal edges
        while dq:
            u = dq.popleft()
            du = dist[u]
            bu = best[u]
            # Traverse neighbors
            for v, eid in g[u]:
                nd = du + 1   # New dist to neighbor
                # Update if better path found
                if nd < dist[v] or (nd == dist[v] and bu < best[v]):
                    dist[v] = nd
                    best[v] = bu
                    dq.append(v)

        # Process queries
        ans = []
        for c in C:
            ans.append("-1" if dist[c] >= INF else str(best[c]))
        out_lines.append(" ".join(ans))
    sys.stdout.write("\n".join(out_lines))


if __name__ == "__main__":
    solve()


# https://github.com/VaHiX/codeForces/

File: Java/ByTier/F/2132_F_Rada_and_the_Chamomile_Valley.java
// Problem: CF 2132 F - Rada and the Chamomile Valley
// https://codeforces.com/contest/2132/problem/F

/*
 * Problem: F. Rada and the Chamomile Valley
 *
 * Purpose:
 *   This solution finds the "essential lanes" (bridges) in a graph from node 1 to node n.
 *   These bridges are lanes that all simple paths between nodes 1 and n must use.
 *   Then, for each query asking which is the nearest essential lane to a given node,
 *   we determine the minimum indexed essential lane within distance 1 in BFS traversal.
 *
 * Algorithms:
 *   - Bridge Finding using Tarjan's Algorithm with DFS (O(n + m))
 *   - Multi-source BFS to compute distances and minimal lane indices (O(n + m))
 *   - Efficient input/output handling with FastReader/PrintWriter
 *
 * Time Complexity: O(n + m + q) per test case
 * Space Complexity: O(n + m) for adjacency list, arrays, etc.
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.*;

public class Main {
    static int n, m;
    static ArrayList<int[]>[] adj; // Adjacency list storing (neighbor, edge_index)
    static int timer;
    static int[] tin, tout, low; // Time in, out, and lowest reachable time
    static ArrayList<Integer> essentialLaneIndices; // Indices of bridges
    static int[] edgeU, edgeV; // Store the endpoints of each edge

    // DFS to find all bridges using Tarjan's algorithm
    static void findEssentialBridgesDfs(int u, int p, int targetNode) {
        tin[u] = low[u] = timer++;
        for (int[] edge : adj[u]) {
            int v = edge[0];         // neighbor node
            int edgeIndex = edge[1]; // index of the current edge
            if (v == p) continue;    // skip parent to avoid cycles

            if (tin[v] != -1) {
                // Back edge found, update low value
                low[u] = Math.min(low[u], tin[v]);
            } else {
                findEssentialBridgesDfs(v, u, targetNode); // DFS on child
                low[u] = Math.min(low[u], low[v]);         // Update low value

                // If this edge is a bridge
                if (low[v] > tin[u]) { 
                    // Check if it belongs to path from start to target node (i.e., in subtree of path)
                    if (tin[v] <= tin[targetNode] && tout[targetNode] <= tout[v]) {
                        essentialLaneIndices.add(edgeIndex);
                    }
                }
            }
        }
        tout[u] = timer++; // Mark finish time
    }

    static void solve() throws IOException {
        n = reader.nextInt();
        m = reader.nextInt();
        adj = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) adj[i] = new ArrayList<>();
        edgeU = new int[m + 1];
        edgeV = new int[m + 1];
        for (int i = 1; i <= m; i++) {
            int u = reader.nextInt();
            int v = reader.nextInt();
            adj[u].add(new int[]{v, i}); // Add bidirectional edges
            adj[v].add(new int[]{u, i});
            edgeU[i] = u;
            edgeV[i] = v;
        }

        tin = new int[n + 1];
        tout = new int[n + 1];
        low = new int[n + 1];
        Arrays.fill(tin, -1); // Initialize all times to -1 (unvisited)
        essentialLaneIndices = new ArrayList<>();
        timer = 0;

        findEssentialBridgesDfs(1, -1, n);

        int q = reader.nextInt();
        if (essentialLaneIndices.isEmpty()) {
            for (int i = 0; i < q; i++) {
                reader.nextInt();
                writer.print(-1);
                if (i < q - 1) writer.print(" ");
            }
            writer.println();
            return;
        }

        // Map each node to the smallest index of an essential lane incident on it
        Map<Integer, Integer> minLaneIndexForNode = new HashMap<>();
        for (int laneIndex : essentialLaneIndices) {
            int u = edgeU[laneIndex];
            int v = edgeV[laneIndex];
            minLaneIndexForNode.put(u, Math.min(minLaneIndexForNode.getOrDefault(u, Integer.MAX_VALUE), laneIndex));
            minLaneIndexForNode.put(v, Math.min(minLaneIndexForNode.getOrDefault(v, Integer.MAX_VALUE), laneIndex));
        }

        // BFS from each critical node to assign distances and minimal lane index
        Set<Integer> criticalNodesSet = minLaneIndexForNode.keySet();
        int[] dist = new int[n + 1];
        int[] ans = new int[n + 1];
        Arrays.fill(dist, -1);
        Arrays.fill(ans, Integer.MAX_VALUE);
        Queue<Integer> qMsbfs = new LinkedList<>();

        // Initialize BFS queue with critical nodes
        for (int node : criticalNodesSet) {
            dist[node] = 0;
            ans[node] = minLaneIndexForNode.get(node);
            qMsbfs.add(node);
        }

        while (!qMsbfs.isEmpty()) {
            int u = qMsbfs.poll();
            for (int[] edge : adj[u]) {
                int v = edge[0];
                if (dist[v] == -1 || dist[v] > dist[u] + 1) {
                    dist[v] = dist[u] + 1;
                    ans[v] = ans[u];
                    qMsbfs.add(v);
                } else if (dist[v] == dist[u] + 1) { 
                    ans[v] = Math.min(ans[v], ans[u]);
                }
            }
        }

        for (int i = 0; i < q; i++) {
            int c = reader.nextInt();
            writer.print(ans[c]);
            if (i < q - 1) writer.print(" ");
        }
        writer.println();
    }

    public static void main(String[] args) throws IOException {
        reader = new FastReader();
        writer = new PrintWriter(System.out);
        int t = reader.nextInt();
        while (t-- > 0) {
            solve();
        }
        writer.close();
    }

    static FastReader reader;
    static PrintWriter writer;

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try { st = new StringTokenizer(br.readLine()); } catch (IOException e) { e.printStackTrace(); }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }
    }
}


// https://github.com/VaHiX/CodeForces/